libr/util/charset.c

```diff
@@ -113,24 +113,36 @@ R_API size_t r_charset_encode_str(RCharset *rc, ut8 *out, size_t out_len, const
 
 // assumes out is as big as in_len
 R_API size_t r_charset_decode_str(RCharset *rc, ut8 *out, size_t out_len, const ut8 *in, size_t in_len) {
-	char k[32];
 	char *o = (char*)out;
-	int i;
-	for (i = 0; i < in_len; i++) {
-		ut8 ch_in = in[i];
-
-		//zero terminate the string
-		snprintf (k, sizeof (k), "%c", ch_in);//snprintf (k, sizeof (k), "0x%02x", ch_in);
-		char *v = sdb_get (rc->db_char_to_hex, k, 0);
-		memmove(v, v+2, strlen (v));
-
-		//convert to ascii
-		char str_hx[32];
-		snprintf (str_hx, sizeof (str_hx), "%c", (char) strtol( v, 0, 16));
-		const char *ret = r_str_get_fail (str_hx, "?");
-		strcpy (o, ret);
-
-		o += strlen (o);
+	int j;
+	ut8 str[320];
+	size_t last_char_size;
+
+	size_t cur;
+	for (cur = 0; cur < in_len; ) {
+		memcpy (str, in, sizeof (str));
+		for (j =( in_len+1 ); j > 0; j--) {
```

Comment:
why in_len+1? why overengineer this loop that can be done forward? for(j=0;j<in_len;j++)

Reply:
because j should be decreasing.

Reply:
yeah i got your idea after sending the review... but instead of using in_len, maybe it will be to compute the maxkeylen and maxkeylen

Reply:
in_len + 1 looks useless and weird. I think `for (j = in_len ; j > 0; j--) {` fits perfectly.

Reply:
No the +1 is necessary.

Reply:
`for (cur = 0; cur < in_len; ) {` sounds better. What do you think?